File: cloudflare/src/decision.mjs
// Port of app/decision.py. Keep parity tests before changing eligibility semantics.
export const ALGORITHM_VERSION = 'decision-1.0-cf1';
export const canonical = value => {
  const text = value.toUpperCase().trim(), iso = text.match(/ISO\s*(?:\/\s*IEC\s*)?(\d{4,5}(?:-\d+)?)/);
  return iso ? `ISO ${iso[1]}` : ({REVENUE:'TURNOVER','ANNUAL TURNOVER':'TURNOVER','PROFESSIONAL INDEMNITY':'INSURANCE'}[text] || text);
};
// Python's round uses ties-to-even.
export function round(x) {const floor=Math.floor(x), fraction=x-floor;return fraction===0.5 ? (floor%2===0 ? floor : floor+1) : Math.round(x);}
const day = value => value?.civilDay || new Date(value).toISOString().slice(0,10);
export function parseDatetime(value) {if(typeof value!=='string'||!/[T ]/.test(value)||!/(Z|[+-]\d\d:\d\d)$/.test(value)||!Number.isFinite(Date.parse(value)))return null;const parsed=new Date(value);parsed.civilDay=value.slice(0,10);return parsed;}
export function evaluate(req,evidence,deadline,today=day(new Date())) {
  const target=deadline?day(deadline):today;
  const base={requirement:req,state:'UNKNOWN',evidence_ids:[],reason:'No confirmed documentary evidence proves this requirement.',next_action:'Add evidence for '+req.capability};
  if(req.not_applicable)return {...base,state:'NOT_APPLICABLE',reason:'Source explicitly states this is not required.',next_action:null};
  if(req.ambiguous||req.capability==='UNKNOWN')return {...base,reason:'Ambiguous or unsupported clause. Review the original source and obtain clarification.',next_action:'Clarify: '+req.text};
  const candidates=evidence.filter(e=>canonical(e.capability)===canonical(req.capability)&&['USER_CONFIRMED','SYSTEM_VERIFIED','EXPIRED'].includes(e.state)&&e.document_id&&e.document_available!==false),failures=[];
  for(const e of candidates){
    if(e.valid_from&&e.valid_from>today)continue;
    if(e.state==='EXPIRED'||(e.valid_until&&e.valid_until<target)){failures.push([e.id,'Evidence expires before the required date; renewal is not assumed.']);continue;}
    const data=e.data||{};
    if(req.edition&&!(data.original_capability||'').includes(req.edition))continue;
    if(req.category==='CERTIFICATION'&&!e.valid_until)continue;
    if(req.currency&&req.currency!==data.currency)continue;
    if(req.threshold!=null){
      let values=[data.value];
      if(req.years){const annual=data.annual_values||{},years=Array.from({length:req.years},(_,i)=>String(Number(target.slice(0,4))-i-1));if(!years.every(y=>y in annual))continue;values=years.map(y=>annual[y]);if(req.period_mode==='average')values=[values.reduce((a,b)=>a+b,0)/values.length];else if(req.period_mode==='aggregate')values=[values.reduce((a,b)=>a+b,0)];}
      if(values.some(v=>v==null))continue;
      const cmp={gte:v=>v>=req.threshold,gt:v=>v>req.threshold,lte:v=>v<=req.threshold}[req.comparator];if(!cmp)continue;
      if(!values.every(cmp)){failures.push([e.id,'Confirmed documentary value does not meet the source threshold.']);continue;}
    }
    return {...base,state:'PASS',reason:'Confirmed document satisfies the normalized predicate and validity window.',evidence_ids:[e.id],next_action:null};
  }
  return failures.length?{...base,state:'FAIL',reason:failures[0][1],evidence_ids:failures.map(x=>x[0])}:base;
}
export function businessDays(start,end){const n=Math.max(0,Math.floor((Date.parse(end)-Date.parse(start))/86400000)),weeks=Math.floor(n/7),rem=n%7;let result=weeks*5;for(let i=0;i<rem;i++){const d=new Date(Date.parse(start)+(weeks*7+i)*86400000).getUTCDay();if(d!==0&&d!==6)result++;}return result;}
const WEIGHTS={technical_fit:.25,evidence_coverage:.3,financial_fit:.15,geography:.1,deadline:.15,capacity:.05};
const ATTR={size_fit:.4,margin:.25,bid_cost_ratio:.2,strategic_fit:.15};
export function analyze(notice,profile,evidence,at=new Date(),lot_id=null){
  const requirements=(notice.requirements||[]).filter(r=>r.lot_id==null||r.lot_id===lot_id),deadline=parseDatetime(notice.deadline),results=requirements.map(r=>evaluate(r,evidence,deadline,day(at)));
  const relevant=results.filter(r=>r.state!=='NOT_APPLICABLE'),weight=r=>r.requirement.hard_gate?3:1,total=relevant.reduce((s,r)=>s+weight(r),0),verified=relevant.filter(r=>r.state==='PASS').reduce((s,r)=>s+weight(r),0),coverage=total?round(verified/total*100):0;
  const failures=results.filter(r=>r.state==='FAIL'&&r.requirement.hard_gate),unknown=results.filter(r=>r.state==='UNKNOWN'&&r.requirement.hard_gate),complete=notice.requirements_complete||false,eligibility=failures.length?'FAIL':unknown.length||!complete?'UNKNOWN':'PASS',friction=[];
  if(notice.country&&!(profile.operating_countries||[]).includes(notice.country))friction.push({reason:'Buyer country is outside declared operating countries. Confirm the place of performance.',source:'buyer_country',kind:'GEOGRAPHY'});
  if(notice.languages?.length&&!notice.languages.some(l=>(profile.languages||[]).includes(l)))friction.push({reason:'The source language is outside declared working languages; confirm submission language and translation needs.',source:'languages',kind:'LANGUAGE'});
  for(const [key,reason] of [['onsite','Onsite delivery is specified.'],['site_visit','A mandatory site visit is specified.']])if(notice[key])friction.push({reason,source:key,kind:'LOGISTICS'});
  const missing=results.filter(r=>['UNKNOWN','FAIL'].includes(r.state)).length,base=12+requirements.length*2+missing*3+(notice.document_count||0)*2+friction.length*5;
  let effort=[round(base*.8),round(base*1.3)];if(profile.effort_multiplier)effort=effort.map(v=>round(v*profile.effort_multiplier));
  const available=deadline?businessDays(day(at),day(deadline))*(profile.bid_hours_per_day??2)*(profile.availability_factor??.8):null;
  const deadlineScore=deadline&&deadline<=at?0:available==null?50:Math.min(100,round(available/Math.max(effort[1],1)*100)),cpvs=notice.cpv_codes||[],interests=profile.cpv_interests||[];
  const overlap=cpvs.filter(c=>interests.some(p=>String(c).slice(0,2)===String(p).slice(0,2))).length,financial=results.filter(r=>r.requirement.category==='FINANCIAL');
  const components={technical_fit:cpvs.length&&interests.length?round(overlap/cpvs.length*100):0,evidence_coverage:coverage,financial_fit:financial.length?round(financial.filter(r=>r.state==='PASS').length/financial.length*100):50,geography:Math.max(0,100-friction.length*25),deadline:deadlineScore,capacity:deadlineScore};
  const feasibility=round(Object.entries(WEIGHTS).reduce((s,[k,w])=>s+components[k]*w,0)),hourly=profile.hourly_cost??60,external=profile.external_bid_cost??0,cost=effort.map(h=>round(h*hourly+external)),value=notice.value,comparable=value!=null&&notice.currency===(profile.currency??'EUR');
  const attr={size_fit:comparable&&value>=(profile.min_contract??0)&&value<=(profile.max_contract??1000000)?100:comparable?20:50,margin:profile.expected_margin!=null?Math.min(100,profile.expected_margin*3):50,bid_cost_ratio:comparable&&value>0?Math.max(0,100-cost[1]/value*1000):50,strategic_fit:profile.strategic_fit??50};
  const attractiveness=round(Object.entries(ATTR).reduce((s,[k,w])=>s+attr[k]*w,0)),confidence=round((notice.completeness_score||0)*.4+coverage*.4+(requirements.length?requirements.reduce((s,r)=>s+r.confidence,0)/requirements.length*100:0)*.2);
  let decision='REVIEW',explanation='Resolve missing or ambiguous evidence before committing bid resources.';
  if(failures.length||notice.status==='CANCELLED'||(deadline&&deadline<=at)){decision='NO_BID';explanation='A mandatory failure, cancellation, or elapsed deadline prevents a supported bid recommendation.';}
  else if(eligibility==='PASS'&&confidence>=75&&deadline&&available>=effort[1]){if(feasibility>=85&&attractiveness>=75){decision='STRONG_BID';explanation='Verified eligibility, strong fit, and sufficient estimated capacity support pursuing this opportunity.';}else if(feasibility>=65&&attractiveness>=50){decision='BID';explanation='Verified eligibility and practical fit support a bid, subject to your final review.';}}
  else if(eligibility==='PASS'&&confidence>=65&&available!=null&&effort[0]<=available&&available<effort[1]){decision='CONDITIONAL_BID';explanation='Eligibility is supported; allocate additional bid capacity before proceeding.';}
  if(available!=null&&available<effort[0]&&decision!=='NO_BID'){decision='REVIEW';explanation='Estimated minimum effort exceeds available bid-team hours.';}
  return {decision,explanation,eligibility,feasibility,attractiveness,confidence,coverage,coverage_numerator:verified,coverage_denominator:total,components:Object.fromEntries(Object.entries(components).map(([k,v])=>[k,{score:v,weight:WEIGHTS[k],basis:k==='evidence_coverage'?'Documentary coverage':'Declared profile and source facts; heuristic'}])),attractiveness_components:Object.fromEntries(Object.entries(attr).map(([k,v])=>[k,{score:round(v),weight:ATTR[k]}])),requirements:results,friction,effort_hours:effort,available_hours:available,cost_range:cost,cost_currency:profile.currency??'EUR',cost_inputs:{hourly_cost:hourly,external_known_cost:external},unmodeled_costs:['Translation, certification, travel and partner costs unless included in your external-cost assumption.'],missing_evidence:results.filter(r=>r.next_action).map(r=>r.next_action),uncertainty:[...complete?[]:['Full tender documentation has not been confirmed complete.'],...deadline?[]:['Authoritative deadline or timezone missing.'],'Competition and incumbency are unavailable unless supported by historical awards.','Effort and attractiveness are estimates, not a win probability.'],sensitivity:['BID','STRONG_BID'].includes(decision)&&available!=null&&available>=effort[1]*1.25?'ROBUST':'FRAGILE',algorithm_version:ALGORITHM_VERSION,model_version:'deterministic',lot_id,analyzed_at:at.toISOString()};
}
export function analyzeLots(notice,profile,evidence,at=new Date()) {const summary=analyze(notice,profile,evidence,at);if(notice.lots?.length){summary.lots=notice.lots.map(l=>analyze({...notice,...l},profile,evidence,at,l.id));summary.decision='REVIEW';summary.explanation='Lot-level assessments govern eligibility. Select the lots you intend to bid for; tender-wide aggregation is not assumed.';}return summary;}

File: app/decision.py
import hashlib
import json
from datetime import UTC, date, datetime, timedelta
from decimal import Decimal

from app.requirements import canonical

ALGORITHM_VERSION = "decision-1.0"
WEIGHTS = {
    "technical_fit": 0.25,
    "evidence_coverage": 0.3,
    "financial_fit": 0.15,
    "geography": 0.1,
    "deadline": 0.15,
    "capacity": 0.05,
}
ATTRACTIVENESS_WEIGHTS = {
    "size_fit": 0.4,
    "margin": 0.25,
    "bid_cost_ratio": 0.2,
    "strategic_fit": 0.15,
}


def parse_datetime(value):
    if not value:
        return None
    if "T" not in value and " " not in value:
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
        return parsed if parsed.tzinfo else None
    except (ValueError, TypeError):
        return None


def cache_key(notice_hash, capability_version, corrections, day=None):
    # A daily time dependency prevents expired evidence/deadlines from retaining a cached PASS.
    return hashlib.sha256(
        json.dumps(
            [
                notice_hash,
                capability_version,
                ALGORITHM_VERSION,
                "deterministic",
                corrections,
                str(day or date.today()),
            ],
            sort_keys=True,
        ).encode()
    ).hexdigest()


def evaluate(req, evidence, deadline, today=None):
    today = today or datetime.now(UTC).date()
    target = deadline.date() if deadline else today
    base = {
        "requirement": req,
        "state": "UNKNOWN",
        "evidence_ids": [],
        "reason": "No confirmed documentary evidence proves this requirement.",
        "next_action": "Add evidence for " + req["capability"],
    }
    if req.get("not_applicable"):
        return base | {
            "state": "NOT_APPLICABLE",
            "reason": "Source explicitly states this is not required.",
            "next_action": None,
        }
    if req.get("ambiguous") or req["capability"] == "UNKNOWN":
        return base | {
            "reason": "Ambiguous or unsupported clause. Review the original source and obtain clarification.",
            "next_action": "Clarify: " + req["text"],
        }
    candidates = [
        e
        for e in evidence
        if canonical(e["capability"]) == canonical(req["capability"])
        and e.get("state") in {"USER_CONFIRMED", "SYSTEM_VERIFIED", "EXPIRED"}
        and e.get("document_id")
        and e.get("document_available", True)
    ]
    failures = []
    for e in candidates:
        if e.get("valid_from") and date.fromisoformat(e["valid_from"]) > today:
            continue
        if e.get("state") == "EXPIRED" or (
            e.get("valid_until") and date.fromisoformat(e["valid_until"]) < target
        ):
            failures.append(
                (e["id"], "Evidence expires before the required date; renewal is not assumed.")
            )
            continue
        data = e.get("data", {})
        if req.get("edition") and req["edition"] not in data.get("original_capability", ""):
            continue
        if req["category"] == "CERTIFICATION" and not e.get("valid_until"):
            continue
        if req.get("currency") and req["currency"] != data.get("currency"):
            continue  # No implicit FX comparison.
        if req.get("threshold") is not None:
            values = [data.get("value")]
            if req.get("years"):
                annual = data.get("annual_values", {})
                required_years = [str(target.year - n) for n in range(1, req["years"] + 1)]
                if not all(year in annual for year in required_years):
                    continue
                values = [annual[year] for year in required_years]
                if req.get("period_mode") == "average":
                    values = [sum(values) / len(values)]
                elif req.get("period_mode") == "aggregate":
                    values = [sum(values)]
            if any(v is None for v in values):
                continue
            threshold = Decimal(str(req["threshold"]))
            comparisons = {
                "gte": lambda v: v >= threshold,
                "gt": lambda v: v > threshold,
                "lte": lambda v: v <= threshold,
            }
            compare = comparisons.get(req.get("comparator"))
            if compare is None:
                continue
            if not all(compare(Decimal(str(v))) for v in values):
                failures.append(
                    (e["id"], "Confirmed documentary value does not meet the source threshold.")
                )
                continue
        return base | {
            "state": "PASS",
            "reason": "Confirmed document satisfies the normalized predicate and validity window.",
            "evidence_ids": [e["id"]],
            "next_action": None,
        }
    if failures:
        return base | {
            "state": "FAIL",
            "reason": failures[0][1],
            "evidence_ids": [v[0] for v in failures],
        }
    return base


def business_days(start, end):
    if end <= start:
        return 0
    total = (end - start).days
    weeks, remainder = divmod(total, 7)
    return weeks * 5 + sum(
        (start + timedelta(days=weeks * 7 + i)).weekday() < 5 for i in range(remainder)
    )


def analyze(notice, profile, evidence, at=None, lot_id=None):
    at = at or datetime.now(UTC)
    requirements = [r for r in notice.get("requirements", []) if r.get("lot_id") in (None, lot_id)]
    deadline = parse_datetime(notice.get("deadline"))
    results = [evaluate(r, evidence, deadline, at.date()) for r in requirements]
    relevant = [r for r in results if r["state"] != "NOT_APPLICABLE"]
    total = sum(3 if r["requirement"]["hard_gate"] else 1 for r in relevant)
    verified = sum(
        3 if r["requirement"]["hard_gate"] else 1 for r in relevant if r["state"] == "PASS"
    )
    coverage = round(verified / total * 100) if total else 0
    failures = [r for r in results if r["state"] == "FAIL" and r["requirement"]["hard_gate"]]
    unknown = [r for r in results if r["state"] == "UNKNOWN" and r["requirement"]["hard_gate"]]
    complete = notice.get("requirements_complete", False)
    eligibility = "FAIL" if failures else "UNKNOWN" if unknown or not complete else "PASS"
    friction = []
    countries = profile.get("operating_countries", [])
    if notice.get("country") and notice["country"] not in countries:
        friction.append(
            {
                "reason": "Buyer country is outside declared operating countries. Confirm the place of performance.",
                "source": "buyer_country",
                "kind": "GEOGRAPHY",
            }
        )
    languages = notice.get("languages", [])
    if languages and not set(languages).intersection(profile.get("languages", [])):
        friction.append(
            {
                "reason": "The source language is outside declared working languages; confirm submission language and translation needs.",
                "source": "languages",
                "kind": "LANGUAGE",
            }
        )
    for key, label in [
        ("onsite", "Onsite delivery is specified."),
        ("site_visit", "A mandatory site visit is specified."),
    ]:
        if notice.get(key):
            friction.append({"reason": label, "source": key, "kind": "LOGISTICS"})
    missing = len([r for r in results if r["state"] in ("UNKNOWN", "FAIL")])
    effort_base = (
        12
        + len(requirements) * 2
        + missing * 3
        + notice.get("document_count", 0) * 2
        + len(friction) * 5
    )
    effort = [round(effort_base * 0.8), round(effort_base * 1.3)]
    if profile.get("effort_multiplier"):
        effort = [round(v * profile["effort_multiplier"]) for v in effort]
    available = (
        business_days(at.date(), deadline.date())
        * profile.get("bid_hours_per_day", 2)
        * profile.get("availability_factor", 0.8)
        if deadline
        else None
    )
    deadline_score = (
        0
        if deadline and deadline <= at
        else 50
        if available is None
        else min(100, round(available / max(effort[1], 1) * 100))
    )
    cpvs, interests = notice.get("cpv_codes", []), profile.get("cpv_interests", [])
    overlap = sum(any(str(c)[:2] == str(p)[:2] for p in interests) for c in cpvs)
    technical = round(overlap / len(cpvs) * 100) if cpvs and interests else 0
    financial_results = [r for r in results if r["requirement"]["category"] == "FINANCIAL"]
    financial = (
        round(sum(r["state"] == "PASS" for r in financial_results) / len(financial_results) * 100)
        if financial_results
        else 50
    )
    components = {
        "technical_fit": technical,
        "evidence_coverage": coverage,
        "financial_fit": financial,
        "geography": max(0, 100 - len(friction) * 25),
        "deadline": deadline_score,
        "capacity": deadline_score,
    }
    feasibility = round(sum(components[k] * w for k, w in WEIGHTS.items()))
    hourly = profile.get("hourly_cost", 60)
    external = profile.get("external_bid_cost", 0)
    cost = [round(h * hourly + external) for h in effort]
    value, currency = notice.get("value"), notice.get("currency")
    comparable = value is not None and currency == profile.get("currency", "EUR")
    minimum, maximum = profile.get("min_contract", 0), profile.get("max_contract", 1000000)
    size_fit = 100 if comparable and minimum <= value <= maximum else 20 if comparable else 50
    margin = profile.get("expected_margin")
    attr = {
        "size_fit": size_fit,
        "margin": min(100, margin * 3) if margin is not None else 50,
        "bid_cost_ratio": max(0, 100 - cost[1] / value * 1000) if comparable and value > 0 else 50,
        "strategic_fit": profile.get("strategic_fit", 50),
    }
    attractiveness = round(sum(attr[k] * w for k, w in ATTRACTIVENESS_WEIGHTS.items()))
    confidence = round(
        (
            notice.get("completeness_score", 0) * 0.4
            + coverage * 0.4
            + (
                sum(r["confidence"] for r in requirements) / len(requirements) * 100
                if requirements
                else 0
            )
            * 0.2
        )
    )
    decision, explanation = (
        "REVIEW",
        "Resolve missing or ambiguous evidence before committing bid resources.",
    )
    if failures or notice.get("status") == "CANCELLED" or (deadline and deadline <= at):
        decision, explanation = (
            "NO_BID",
            "A mandatory failure, cancellation, or elapsed deadline prevents a supported bid recommendation.",
        )
    elif eligibility == "PASS" and confidence >= 75 and deadline and available >= effort[1]:
        if feasibility >= 85 and attractiveness >= 75:
            decision, explanation = (
                "STRONG_BID",
                "Verified eligibility, strong fit, and sufficient estimated capacity support pursuing this opportunity.",
            )
        elif feasibility >= 65 and attractiveness >= 50:
            decision, explanation = (
                "BID",
                "Verified eligibility and practical fit support a bid, subject to your final review.",
            )
    elif (
        eligibility == "PASS"
        and confidence >= 65
        and available is not None
        and effort[0] <= available < effort[1]
    ):
        decision, explanation = (
            "CONDITIONAL_BID",
            "Eligibility is supported; allocate additional bid capacity before proceeding.",
        )
    if available is not None and available < effort[0] and decision != "NO_BID":
        decision, explanation = (
            "REVIEW",
            "Estimated minimum effort exceeds available bid-team hours.",
        )
    return {
        "decision": decision,
        "explanation": explanation,
        "eligibility": eligibility,
        "feasibility": feasibility,
        "attractiveness": attractiveness,
        "confidence": confidence,
        "coverage": coverage,
        "coverage_numerator": verified,
        "coverage_denominator": total,
        "components": {
            k: {
                "score": v,
                "weight": WEIGHTS[k],
                "basis": "Documentary coverage"
                if k == "evidence_coverage"
                else "Declared profile and source facts; heuristic",
            }
            for k, v in components.items()
        },
        "attractiveness_components": {
            k: {"score": round(v), "weight": ATTRACTIVENESS_WEIGHTS[k]} for k, v in attr.items()
        },
        "requirements": results,
        "friction": friction,
        "effort_hours": effort,
        "available_hours": available,
        "cost_range": cost,
        "cost_currency": profile.get("currency", "EUR"),
        "cost_inputs": {"hourly_cost": hourly, "external_known_cost": external},
        "unmodeled_costs": [
            "Translation, certification, travel and partner costs unless included in your external-cost assumption."
        ],
        "missing_evidence": [r["next_action"] for r in results if r["next_action"]],
        "uncertainty": (
            [] if complete else ["Full tender documentation has not been confirmed complete."]
        )
        + ([] if deadline else ["Authoritative deadline or timezone missing."])
        + [
            "Competition and incumbency are unavailable unless supported by historical awards.",
            "Effort and attractiveness are estimates, not a win probability.",
        ],
        "sensitivity": "ROBUST"
        if decision in ("BID", "STRONG_BID")
        and available is not None
        and available >= effort[1] * 1.25
        else "FRAGILE",
        "algorithm_version": ALGORITHM_VERSION,
        "model_version": "deterministic",
        "lot_id": lot_id,
        "analyzed_at": at.isoformat(),
    }


def analyze_lots(notice, profile, evidence, at=None):
    lots = notice.get("lots", [])
    if not lots:
        return analyze(notice, profile, evidence, at)
    results = [analyze(notice | lot, profile, evidence, at, lot["id"]) for lot in lots]
    summary = analyze(notice, profile, evidence, at)
    summary["lots"] = results
    summary["decision"] = "REVIEW"
    summary["explanation"] = (
        "Lot-level assessments govern eligibility. Select the lots you intend to bid for; tender-wide aggregation is not assumed."
    )
    return summary
